packages/core/src/GondelDomUtils.ts

```diff
@@ -50,69 +93,104 @@ export function stopComponents(domContext?: ArrayLikeHtmlElement, namespace: str
 }
 
 /**
  * Returns the gondel instance for the given HtmlELement once it is booted
  */
-export function getComponentByDomNodeAsync(
+export function getComponentByDomNodeAsync<T extends GondelComponent>(
   domNode: ArrayLikeHtmlElement,
   namespace: string = "g"
-): Promise<GondelComponent> {
+): Promise<T> {
   const firstNode = getFirstDomNode(domNode);
-  const gondelComponent = (firstNode as any)[`_gondelA_${namespace}`];
+  const gondelComponent = (firstNode as any)[getGondelAttribute(namespace, "async")];
   // Stop if this dom node is not known to gondel
   if (!gondelComponent) {
     return Promise.reject(undefined);
   }
   // or the component is already booted up return it
   if (gondelComponent._ctx) {
-    return Promise.resolve(gondelComponent as GondelComponent);
+    return Promise.resolve(gondelComponent as T);
   }
   // Wait the component to boot up and return it
-  return gondelComponent.then(() => (firstNode as any)[`_gondel_${namespace}`]);
+  return gondelComponent.then(() => (firstNode as any)[getGondelAttribute(namespace)]);
 }
 
 /**
  * Returns all components inside the given node
  */
+
 export function findComponents(
+  domNode?: ArrayLikeHtmlElement,
+  component?: string,
+  namespace?: string
+): Array<GondelComponent>;
+export function findComponents<T extends GondelComponent>(
+  domNode?: ArrayLikeHtmlElement,
+  component?: { new (): T },
+  namespace?: string
+): Array<T>;
+export function findComponents<T extends GondelComponent>(
   domNode: ArrayLikeHtmlElement = document.documentElement,
-  componentName?: string,
+  component?: { new (): T } | string,
   namespace: string = "g"
-): Array<GondelComponent> {
+): Array<T | GondelComponent> {
   const firstNode = getFirstDomNode(domNode);
-  const components: Array<GondelComponent> = [];
-  const attribute = `_gondel_${namespace}`;
+  const components: Array<T | GondelComponent> = [];
+  const attribute = getGondelAttribute(namespace);
   const nodes = firstNode.querySelectorAll(
-    `[data-${namespace}-name${componentName ? `="${componentName}"` : ""}]`
+    `[data-${namespace}-name${
+      component
+        ? `="${
+            typeof component === "string"
+              ? component
+              : getComponentName(component as IGondelComponent<T>, namespace)
+          }"`
+        : ""
+    }]`
   );
   for (let i = 0; i < nodes.length; i++) {
     const node = nodes[i];
-    const gondelComponentInstance = (node as any)[attribute] as GondelComponent;
+    const gondelComponentInstance = (node as any)[attribute] as T & GondelComponent;
     // Verify that the component is fully booted
     if (gondelComponentInstance && gondelComponentInstance._ctx === node) {
       components.push(gondelComponentInstance);
     }
   }
-  return components;
+
+  if (typeof component === "string") {
+    return components as Array<GondelComponent>;
+  }
+
+  return components as Array<T & GondelComponent>;
```

Comment:
Do we really need ` & GondelComponent` here?

Reply:
Yes, I checked how the behaviour would be if we remove GondelComponent, and it looks like that TypeScript won't resolve the BaseComponent type so you wouldn't have access to `_ctx` etc.